src/services/kratos/index.ts

```diff
@@ -0,0 +1,194 @@
+import { getKratosConfig, getKratosMasterPhonePassword } from "@config"
+import {
+  AuthenticationKratosError,
+  IncompatibleSchemaUpgradeError,
+  KratosError,
+  LikelyNoUserWithThisPhoneExistError,
+  LikelyUserAlreadyExistError,
+  UnknownKratosError,
+} from "@domain/users/errors"
+import {
+  AdminUpdateIdentityBody,
+  Configuration,
+  Identity,
+  Session,
+  SuccessfulSelfServiceLoginWithoutBrowser,
+  SuccessfulSelfServiceRegistrationWithoutBrowser,
+  V0alpha2Api,
+  V0alpha2ApiInterface,
+} from "@ory/client"
+
+import { AxiosResponse } from "node_modules/@ory/client/node_modules/axios/index"
+
+const { publicApi, adminApi } = getKratosConfig()
+
+const KratosSdk: (kratosEndpoint: string) => V0alpha2ApiInterface = (kratosEndpoint) =>
+  new V0alpha2Api(new Configuration({ basePath: kratosEndpoint }))
+
+export const kratosPublic = KratosSdk(publicApi)
+export const kratosAdmin = KratosSdk(adminApi)
+
+// login with phone
+const password = getKratosMasterPhonePassword()
+
+export const loginForPhoneNoPasswordSchema = async (
+  phone: PhoneNumber,
+): Promise<LoginForPhoneNoPasswordSchemaResponse | KratosError> => {
+  const flow = await kratosPublic.initializeSelfServiceLoginFlowWithoutBrowser()
+
+  const identifier = phone
+  const method = "password"
+
+  let result: AxiosResponse<SuccessfulSelfServiceLoginWithoutBrowser>
+
+  try {
+    result = await kratosPublic.submitSelfServiceLoginFlow(flow.data.id, {
+      identifier,
+      method,
+      password,
+    })
+  } catch (err) {
+    if (err.message === "Request failed with status code 400") {
+      return new LikelyNoUserWithThisPhoneExistError(err)
+    }
+
+    if (err.message === "Request failed with status code 401") {
+      return new AuthenticationKratosError(err)
+    }
+
+    return new UnknownKratosError(err)
+  }
+
+  const sessionToken = result.data.session_token as KratosSessionToken
+  const kratosUserId = result.data.session.identity.id as KratosUserId
+  return { sessionToken, kratosUserId }
+}
+
+export const createKratosUserForPhoneNoPasswordSchema = async (
+  phone: PhoneNumber,
+): Promise<CreateKratosUserForPhoneNoPasswordSchemaResponse | KratosError> => {
+  const flow = await kratosPublic.initializeSelfServiceRegistrationFlowWithoutBrowser()
+
+  const traits = { phone }
+  const method = "password"
+
+  let result: AxiosResponse<SuccessfulSelfServiceRegistrationWithoutBrowser>
+
+  try {
+    result = await kratosPublic.submitSelfServiceRegistrationFlow(flow.data.id, {
+      traits,
+      method,
+      password,
+    })
+  } catch (err) {
+    if (err.message === "Request failed with status code 400") {
+      return new LikelyUserAlreadyExistError(err)
```

Comment:
Why `Likely`?

Reply:
I'm using the public API here. this is the only way to get a session token. so the api doesn't return much context on the error by design - because it's public. 

so the error can be that it's exist, but something else also. the only way to know for sure is to look at the logs.